### PR TITLE
Ensure well-formed and concrete types before IR gen

### PR DESCRIPTION
**Pull request by sezna**
The IR stage is written such that any unresolved types are an internal compiler error. This is an invariant of the IR according to @otrho and therefore should be upheld as an ICE.

This PR adds a pass before the IR stage that ensures all types are well-formed and not insufficiently specified before IR gen. Eventually we could make a _new_ tree type that uses a specified `Type` enum without any unspecified variants. That would get us compile-time proofs of correctness and allow us to drop the type engine after type checking. However, it would be a _ton_ of copy/paste as we would have to recreate our AST type, so I'm deeming it out of scope for now.

before:
image

after:
image

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_fail/insufficient_type_info/Forc.toml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[project]
+authors = ["Sezna"]
```
Comment (adlerjohn) [inline test/src/e2e_vm_tests/test_programs/should_fail/insufficient_type_info/Forc.toml:2]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

**Review by adlerjohn — CHANGES_REQUESTED** [review]

File: sway-core/src/semantic_analysis/ast_node/mod.rs [hunk 2]
```diff
@@ -94,6 +94,35 @@ impl std::fmt::Display for TypedAstNode {
...
+    /// the [TreeType] is Script or Predicate.
+    pub(crate) fn is_main_function(&self, tree_type: TreeType) -> bool {
+        match &self {
+            TypedAstNode {
+                content:
+                    TypedAstNodeContent::Declaration(TypedDeclaration::FunctionDeclaration(
+                        TypedFunctionDeclaration { name, .. },
+                    )),
+                ..
+            } if name.as_str() == "main" => {
```
Comment (mohammadfawaz) [inline sway-core/src/semantic_analysis/ast_node/mod.rs:120]: Can the string `"main"` be moved the file that contains all the constants? I'm going to assume that this string is used in multiple places throughout the codebase.

File: test/src/e2e_vm_tests/test_programs/should_fail/insufficient_type_info/src/main.sw [hunk 3]
```diff
@@ -0,0 +1,10 @@
+script;
+
+
+fn foo<T>() {
+  let x = size_of::<T>();
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_fail/insufficient_type_info/src/main.sw:5]: Formatter :)

File: sway-core/src/type_engine/unresolved_type_check.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                        .map(|x| &x.type_id)
+                        .flat_map(UnresolvedTypeCheck::check_for_unresolved_types)
+                        .collect(),
+                );
+                body
+            }
+            ConstantDeclaration(TypedConstantDeclaration { value, .. }) => {
+                value.check_for_unresolved_types()
+            }
+            StorageReassignment(TypeCheckedStorageReassignment { fields, rhs, .. }) => fields
```
Comment (mohammadfawaz) [inline sway-core/src/type_engine/unresolved_type_check.rs:232]: `StorageReassignment` cannot have generics either. I suppose this doesn't hurt though.

File: sway-core/src/type_engine/unresolved_type_check.rs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+/// error to signal to the user that more type information is needed.
+pub(crate) trait UnresolvedTypeCheck {
+    fn check_for_unresolved_types(&self) -> Vec<CompileError>;
+}
+
+impl UnresolvedTypeCheck for TypeId {
+    fn check_for_unresolved_types(&self) -> Vec<CompileError> {
+        use TypeInfo::*;
+        match look_up_type_id(*self) {
+            UnknownGeneric { name } => vec![CompileError::UnableToInferGeneric {
```
Comment (mohammadfawaz) [inline sway-core/src/type_engine/unresolved_type_check.rs:18]: Small nit regarding how this error is defined in `errors.rs`. Should we change it to be closer to Rust? Something like: 
"cannot infer type for type parameter `T`..."

**Comment by mohammadfawaz** [comment]
I like this change very much. A verification pass for each major step in the compiler helps detecting and debugging errors. Also helps achieve a #beyondmonolithic compiler 😁

File: sway-core/src/type_engine/unresolved_type_check.rs (hunk 4 above)
Comment (sezna) [reply sway-core/src/type_engine/unresolved_type_check.rs:232]: It was easy enough to write...maybe we will have generic storage in the future...shrug

**Review by mohammadfawaz — APPROVED** [review]
LGTM!

**Comment by mohammadfawaz** [comment]
Wait, there's a 14 MB binary files added in this change called `nvim.appimage` 😨

**Comment by otrho** [comment]
I'll all for Neovim advocacy but this is a bit much. 😉

**Review by otrho — COMMENTED** [review]
That was more involved that I thought it would be.

If you're up for it, while removing the appimage you could also patch this to use the 'main' constant:
https://github.com/FuelLabs/sway/blob/95816e4e41aae1d3425ba6ff5e7266076d8400fa/sway-core/src/lib.rs#L481

Otherwise, looks good.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**